ddtrace/internal/encoding.py

```diff
@@ -48,17 +53,36 @@ def encode(self, obj):
         raise NotImplementedError()
 
 
-class JSONEncoder(_EncoderBase):
+class JSONEncoder(json.JSONEncoder, _EncoderBase):
     content_type = "application/json"
 
+    def __init__(self):
+        # Reduce unnecessary whitespace
+        super(JSONEncoder, self).__init__(separators=(",", ":"), indent=None)
+
     def encode_traces(self, traces):
-        normalized_traces = [[span.to_dict() for span in trace] for trace in traces]
+        normalized_traces = [[JSONEncoder._normalize_span(span.to_dict()) for span in trace] for trace in traces]
         return self.encode(normalized_traces)
 
     @staticmethod
-    def encode(obj):
-        # type: (Any) -> str
-        return json.dumps(obj)
+    def _normalize_span(span):
+        # Ensure all string attributes are actually strings and not bytes
+        span["resource"] = JSONEncoder._normalize_str(span["resource"])
+        span["name"] = JSONEncoder._normalize_str(span["name"])
+        span["service"] = JSONEncoder._normalize_str(span["service"])
+        # TODO: What about meta?
+        return span
+
+    @staticmethod
+    def _normalize_str(obj):
+        if PY3:
+            if isinstance(obj, (binary_type, text_type)):
+                return ensure_text(obj, errors="backslashreplace")
```

Comment:
Could we not just use this in all cases?

Reply:
No, because if `obj` is not a string type, then we want to call `stringify()` below.


It _should_ be a string type, so depends on whether we want to be defensive here?

Reply:
> It should be a string type, so depends on whether we want to be defensive here?

Hmm. At this point I think we should assume that we have "good data", according to our API at least else we're duplicating the logic of `stringify`-ing done in Span, permitting broken usage, and creating inconsistency:

```python
span = tracer.trace(...)
span.name = object()  # should be a programming error/type violation
span.finish()
``` 

Results in a TypeError with the msgpack encoder https://github.com/DataDog/dd-trace-py/blob/e10a574e9d1f11ca4037ea0b37e3723f8ab9c99f/ddtrace/internal/_encoding.pyx#L142 

but would work with the JSONEncoder with this change.